### PR TITLE
Refactoring/retry analyzer

### LINKED ISSUE #65: Improvement and refactoring of Retry analyzer
**Is your feature request related to a problem? Please describe.**
The current implementation of Testerra Retry analyzer only allows to add custom rules. The current default rule `Timeout exception` does not work proberly and it is also not a retry cause.

**Describe the solution you'd like**
We decied to define the following exceptions as a retry cause:
* `org.openqa.selenium.remote.UnreachableBrowserException`
* `org.openqa.selenium.json.JsonException`
* `org.openqa.selenium.json.WebDriverException`

These retry rules are active by default.

The user can add additional retry causes according the current implementation.

### PR DESCRIPTION
**Pull request by mreiche**
# Description

* This change improves the `RetryAnalyzer` API to make it more clean and slightly faster.
* Added `SeleniumRetryAnalyzer` for typical Selenium exceptions in `driver-ui-desktop`
* Changed the interface to `Optional` of the `AdditionalRetryAnalyzer` to prevent Nullpointers.
* Mark out `Retry Failed` test cases in `report-ng`
image
image

Fixes #65

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/hooks/DriverUiHookDesktop.java [hunk 1]
```diff
@@ -43,6 +45,7 @@
     public void init() {
         WebDriverManager.registerWebDriverFactory(new DesktopWebDriverFactory(), browsers);
         GuiElement.registerGuiElementCoreFactory(new DesktopGuiElementCoreFactory(), browsers);
+        RetryAnalyzer.registerAdditionalRetryAnalyzer(new SeleniumRetryAnalyzer());
```
Comment (martingrossmann) [inline driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/hooks/DriverUiHookDesktop.java:48]: Whats about the WebDriverRetryAnalyzer? It should also a default.

File: driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/execution/testng/SeleniumRetryAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * under the License.
+ */
+
+package eu.tsystems.mms.tic.testframework.execution.testng;
+
+import java.util.Optional;
+import org.openqa.selenium.json.JsonException;
+import org.openqa.selenium.remote.UnreachableBrowserException;
+
+public class SeleniumRetryAnalyzer implements AdditionalRetryAnalyzer {
```
Comment (martingrossmann) [inline driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/execution/testng/SeleniumRetryAnalyzer.java:28]: The documentation needs an update about this new default retry analyzer and the WebDriver retry analyzer. The file `https://github.com/telekom/testerra/blob/master/docs/src/docs/execution/execution-retry-analyzer.adoc` in `Specific retries` contains very old stuff from a former framework.

**Review by martingrossmann — CHANGES_REQUESTED** [review]

File: driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/hooks/DriverUiHookDesktop.java (hunk 1 above)
Comment (mreiche) [reply driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/hooks/DriverUiHookDesktop.java:48]: https://github.com/telekom/testerra/blob/refactoring/retry-analyzer/driver-ui/src/main/java/eu/tsystems/mms/tic/testframework/hooks/DriverUiHook.java#L76

File: driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/execution/testng/SeleniumRetryAnalyzer.java (hunk 2 above)
Comment (martingrossmann) [reply driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/execution/testng/SeleniumRetryAnalyzer.java:28]: Line 52ff still contains some old stuff. Its located in the extract of the `test.properties`

Comment (martingrossmann) [reply driver-ui-desktop/src/main/java/eu/tsystems/mms/tic/testframework/execution/testng/SeleniumRetryAnalyzer.java:28]: Ok, I recognized the `WebDriverRetryAnalyzer` still contains message checks, so there wasn't so much old stuff. Sorry for confusing.

**Comment by martingrossmann** [comment]
The `WebDriverRetryAnalyzer` is triggered if a `WebDriverException` occurs and if it contains specific messages. But it seems there is a duplicate:
The `UnreachableBrowserException` extends from `WebDriverException` and occues with `"Error communicating with the remote browser. It may have died."` ( https://github.com/SeleniumHQ/selenium/blob/trunk/java/client/src/org/openqa/selenium/remote/RemoteWebDriver.java)
We should remove the message check from `WebDriverRetryAnalyzer`.

**Comment by martingrossmann** [comment]
@gogro The exception `JsonException` can be thrown in different ways with many messages: https://github.com/SeleniumHQ/selenium/search?q=throw+new+JsonException
Can you specify which detailed messages mostly occurs for that exception?

**Comment by gogro** [comment]
> @gogro The exception `JsonException` can be thrown in different ways with many messages: https://github.com/SeleniumHQ/selenium/search?q=throw+new+JsonException
> Can you specify which detailed messages mostly occurs for that exception?

only this error string :
org.openqa.selenium.json.JsonException: Expected to read a START_MAP but instead have: END.

happens on idle timeout iirc

**Comment by martingrossmann** [comment]
> > @gogro The exception `JsonException` can be thrown in different ways with many messages: https://github.com/SeleniumHQ/selenium/search?q=throw+new+JsonException
> > Can you specify which detailed messages mostly occurs for that exception?
> 
> only this error string :
> org.openqa.selenium.json.JsonException: Expected to read a START_MAP but instead have: END.
> 
> happens on idle timeout iirc

@mreiche Please, can you add this message to specify the `JsonException`?

**Review by martingrossmann — APPROVED** [review]

**Review by erickubenka — APPROVED** [review]

**Pull request merged**